### DISCUSSION TITLE
Workaround/support oft non-integer major/minor/patch-versions?

### DISCUSSION THREAD

**Discussion by thomaszub** [renovatebot/renovate#19447] [Request Help]
Hi,

I'm currently stumbling on date-based dependency versions and the correct separation into major, minor, etc.

For example take two versions of the same dependency:
`2022.09.5` -> `2022.09` as major, `5` as minor (no patch)
`2022.12.1` -> `2022.12` as major, `1` as minor (no patch)

I tried the following versioning regex:
`"versioning": "regex:^(?<major>\\d+\\.\\d+)\\.(?<minor>\\d+)(?:\\.(?<patch>\\d+))?$"`

The regex itself is working, but the version parts are converted to integers stripping away the second part
`2022.09.5` -> `2022` as major, `5` as minor (no patch)
`2022.12.1` -> `2022` as major, `1` as minor (no patch)

Do you know a workaround for this kind of versioning or do you plan to support non-integer version parts?

Kind regards
Thomas

**Comment by rarkins** [comment]
I'm not sure we have ever been asked about decimal major/minor/patch parts before. I think in this case you'd be fine if you treated the major as `202209`, `202212`, etc, but I'm not sure how to do that unless you used the regex manager as well and constructed the `currentValue` using `currentValueTemplate`.

**Reply by thomaszub** [reply]
Thanks for the suggestion, @rarkins! 

Do you have a short example with `currentvalueTemplate` as I could not find any in the documentation?
As far as I understand, there are multiple problems with a regex manager working with Maven pom.xml:
1. The `matchStrings` works on the source file, making it hard to write a correct matcher as Maven dependencies are XML elements without a fixed order between `groupId`, `artifactId` and `version`. Additionally a `version` can be set by a property.
2. The `currentValue` in a `currentValueTemplate` can be extracted by a regex with a named capture group `currentValue` (just an assumption, could not find any example). Unfortunately this leads to the same problem. RE2 can not remove the dot with a single regex.

Two possible alternatives I thought of:
1. Implementation of https://github.com/renovatebot/renovate/issues/17340
2. The possibility to add custom JavaScript implementing the Versioning-API
3. Support for decimal versions. From JavaScript/TypeScript perspective this should be a minor change as both are type `number`. I could make a PR if you are interested in this change.

Alternative 1 sounds promising to me. Would you accept such an extension?

**Reply by HonkingGoose** [reply]
> Do you have a short example with `currentvalueTemplate` as I could not find any in the documentation?

I found an example which uses `currentvalueTemplate` in our docs, but it's a bit hidden: [^quote]

```json
{
  "regexManagers": [
    {
      "fileMatch": ["^versions.ini$"],
      "matchStrings": ["GOOGLE_API_VERSION=(?<currentDigest>.*?)\\n"],
      "currentValueTemplate": "master",
      "depNameTemplate": "googleapis",
      "packageNameTemplate": "https://github.com/googleapis/googleapis",
      "datasourceTemplate": "git-refs"
    }
  ]
}
```

[^quote]: https://docs.renovatebot.com/modules/datasource/#git-refs-datasource

**Reply by rarkins** [reply]
This one has an example of `currentValueTemplate` being templated: https://github.com/renovatebot/renovate/discussions/17342

So it might work if it's `"{{year}}{{month}}.{{day}}"` and you extract the others from `matchStrings`
